cleanlab/datalab/issue_manager/noniid.py

```diff
@@ -0,0 +1,452 @@
+        The KS statistic between the two ECDFs.
+
+    Note
+    ----
+    - Both input arrays should have the same length.
+    - The input arrays are histograms, which means they contain the count
+      or frequency of values in each group. The data in the histograms
+      should be normalized so that they sum to one.
+
+    To calculate the KS statistic, the function first calculates the ECDFs
+    for both input arrays, which are step functions that show the cumulative
+    sum of the data up to each point. The function then calculates the
+    largest absolute difference between the two ECDFs.
+    """
+
+    neighbor_cdf = np.cumsum(neighbor_histogram)
+    non_neighbor_cdf = np.cumsum(non_neighbor_histogram)
+
+    statistic = np.max(np.abs(neighbor_cdf - non_neighbor_cdf))
+    return statistic
+
+
+class NonIIDIssueManager(IssueManager):
+    """Manages issues related to non-iid data distributions.
+
+    Parameters
+    ----------
+    datalab :
+        The Datalab instance that this issue manager searches for issues in.
+
+    metric :
+        The distance metric used to compute the KNN graph of the examples in the dataset.
+        If set to `None`, the metric will be automatically selected based on the dimensionality
+        of the features used to represent the examples in the dataset.
+
+    k :
+        The number of nearest neighbors to consider when computing the KNN graph of the examples.
+
+    num_permutations :
+        The number of trials to run when performing permutation testing to determine whether
+        the distribution of index-distances between neighbors in the dataset is IID or not.
+
+    """
+
+    description: ClassVar[
+        str
+    ] = """The NonIIDIssueManager detects whether the given dataset is sampled IID or not.
+
+    Data can be non-IID in many ways and in general it is impossible
+    to detect all cases of non-IID sampling.  This issue manager
+    investigates whether the ordering of examples in the dataset is
+    dependent on whether examples are neighbors in the KNN graph or
+    not. The algorithm uses permutation testing with the
+    Kolmogorov-Smirnov statistic to determine whether the distribution
+    of index-distances between neighbors in the dataset is
+    significantly different than that of the distribution of all
+    index-distances in the dataset.
+
+    Detecting non-IID data can very important when collecting datasets
+    or preparing a model for deployment.  Although shuffling data is
+    generally good practice for removing non-IID issues, knowing that
+    there are underlying problems with distribution drift or dependent
+    sampling during data collection is important to know in order to
+    understand the real-world environment that your model will be
+    deployed in.
+
+    Types of non-IID problems in datasets can be:
+        - Distribution drift or concept drift
+        - Video frames in an image dataset
+        - Sorting
+        - Dependent sampling
```

Comment:
```suggestion
    ] = """Whether the dataset exhibits statistically significant
          violations of the IID assumption like: 
          changepoints or shift, drift, autocorrelation, etc.
          The specific violation considered is whether the 
          examples are ordered such that almost adjacent examples
          tend to have more similar feature values.
```

Reply:
save the rest of description for docs page dedicated to describing all issue types. 
Add note to the rest of description that we should mention shuffling and why it is not a fix for this issue